In [2]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense
from tensorflow.keras.utils import to_categorical

# Constants
SEQUENCE_LENGTH = 40
BATCH_SIZE = 128
EPOCHS = 50


In [3]:
# Sample data
data = """
Two households, both alike in dignity,
In fair Verona, where we lay our scene,
From ancient grudge break to new mutiny,
Where civil blood makes civil hands unclean.
From forth the fatal loins of these two foes
A pair of star-cross'd lovers take their life;
Whose misadventured piteous overthrows
Do with their death bury their parents' strife.
The fearful passage of their death-mark'd love,
And the continuance of their parents' rage,
Which, but their children's end, nought could remove,
Is now the two hours' traffic of our stage;
The which if you with patient ears attend,
What here shall miss, our toil shall strive to mend.
"""

# Convert all characters to lowercase
data = data.lower()


In [4]:
chars = sorted(list(set(data)))
VOCAB_SIZE = len(chars)


In [5]:
char_to_int = {c: i for i, c in enumerate(chars)}
int_to_char = {i: c for i, c in enumerate(chars)}

# Convert the character sequence input into an integer sequence
int_data = [char_to_int[char] for char in data]


In [6]:
model = Sequential([
    Embedding(VOCAB_SIZE, 256, input_length=SEQUENCE_LENGTH),
    LSTM(512, return_sequences=True),
    LSTM(512),
    Dense(VOCAB_SIZE, activation='softmax')
])


2023-10-08 07:55:30.877674: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node
Your kernel may have been built without NUMA support.
2023-10-08 07:55:30.878823: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1956] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2023-10-08 07:55:31.580919: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_2_grad/concat/split_2/split_dim' with dtype int32
	 [[{{node gradients/split_2_grad/conca

In [7]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])


In [8]:
X = []
Y = []

for i in range(0, len(int_data) - SEQUENCE_LENGTH, 1):
    sequence_in = int_data[i:i + SEQUENCE_LENGTH]
    sequence_out = int_data[i + SEQUENCE_LENGTH]
    X.append(sequence_in)
    Y.append(sequence_out)

X = np.array(X)
Y = to_categorical(Y)


In [9]:
model.fit(X, Y, batch_size=BATCH_SIZE, epochs=EPOCHS)


Epoch 1/50


2023-10-08 07:56:36.796462: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_2_grad/concat/split_2/split_dim' with dtype int32
	 [[{{node gradients/split_2_grad/concat/split_2/split_dim}}]]
2023-10-08 07:56:36.798954: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'gradients/split_grad/concat/split/split_dim' with dtype int32
	 [[{{node gradients/split_grad/concat/split/split_dim}}]]
2023-10-08 07:56:36.801097: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You mus

5/5 [==============================] - 7s 717ms/step - loss: 3.3022 - accuracy: 0.1100
Epoch 2/50
5/5 [==============================] - 3s 625ms/step - loss: 3.1566 - accuracy: 0.0981
Epoch 3/50
5/5 [==============================] - 3s 630ms/step - loss: 3.0201 - accuracy: 0.1472
Epoch 4/50
5/5 [==============================] - 3s 634ms/step - loss: 3.0208 - accuracy: 0.1421
Epoch 5/50
5/5 [==============================] - 3s 649ms/step - loss: 2.9982 - accuracy: 0.1472
Epoch 6/50
5/5 [==============================] - 3s 610ms/step - loss: 2.9928 - accuracy: 0.1472
Epoch 7/50
5/5 [==============================] - 3s 598ms/step - loss: 2.9923 - accuracy: 0.1472
Epoch 8/50
5/5 [==============================] - 3s 607ms/step - loss: 2.9869 - accuracy: 0.1472
Epoch 9/50
5/5 [==============================] - 3s 619ms/step - loss: 2.9769 - accuracy: 0.1472
Epoch 10/50
5/5 [==============================] - 3s 620ms/step - loss: 2.9709 - accuracy: 0.1472
Epoch 11/50
5/5 [=============

#Output
"""
5/5 [==============================] - 7s 717ms/step - loss: 3.3022 - accuracy: 0.1100
Epoch 2/50
5/5 [==============================] - 3s 625ms/step - loss: 3.1566 - accuracy: 0.0981
Epoch 3/50
5/5 [==============================] - 3s 630ms/step - loss: 3.0201 - accuracy: 0.1472
Epoch 4/50
5/5 [==============================] - 3s 634ms/step - loss: 3.0208 - accuracy: 0.1421
Epoch 5/50
5/5 [==============================] - 3s 649ms/step - loss: 2.9982 - accuracy: 0.1472
Epoch 6/50
5/5 [==============================] - 3s 610ms/step - loss: 2.9928 - accuracy: 0.1472
Epoch 7/50
5/5 [==============================] - 3s 598ms/step - loss: 2.9923 - accuracy: 0.1472
Epoch 8/50
5/5 [==============================] - 3s 607ms/step - loss: 2.9869 - accuracy: 0.1472
Epoch 9/50
5/5 [==============================] - 3s 619ms/step - loss: 2.9769 - accuracy: 0.1472
Epoch 10/50
5/5 [==============================] - 3s 620ms/step - loss: 2.9709 - accuracy: 0.1472
Epoch 11/50
5/5 [==============================] - 3s 626ms/step - loss: 2.9589 - accuracy: 0.1472
Epoch 12/50
5/5 [==============================] - 3s 628ms/step - loss: 2.9457 - accuracy: 0.1472
Epoch 13/50
5/5 [==============================] - 3s 623ms/step - loss: 2.9096 - accuracy: 0.1472
...
Epoch 49/50
5/5 [==============================] - 3s 626ms/step - loss: 0.0710 - accuracy: 0.9949
Epoch 50/50
5/5 [==============================] - 3s 627ms/step - loss: 0.0575 - accuracy: 1.0000
"""